src/handlers/canonicalFieldsHandlers.ts

```diff
@@ -33,6 +45,75 @@ const getCanonicalFields = (req: Request, res: Response): void => {
     });
 };
 
+const postCanonicalFieldBodySchema: JSONSchemaType<Omit<CanonicalField, 'createdAt' | 'id'>> = {
```

Comment:
This is redundant -- I've opened [an issue in the ajv repository](https://github.com/ajv-validator/ajv/issues/2085) to explore the idea of creating corresponding "utility" functions in the world of ajv to mirror the TS utility types.

In the mean time I think it will be valuable to create those utility methods directly in our repo and use them, so I'll go add that in.

Reply:
I spent a time boxed two hours trying to create a utility that would accomplish this goal, but it's... incredibly complex (maybe impossible) to do in a type safe way.  Here's where I got (this doesn't work but since I'm not gonna commit it I'd like to post it SOMEWHERE for posterity)

```

const removeAttributesFromObject = <T, K extends string | number | symbol>(
  obj: T,
  keysToRemove: (keyof T)[],
): Omit<T, K> => {
  const modifiedObj: Partial<T> = { ...obj };
  keysToRemove.forEach((key) => {
    modifiedObj[key] = undefined;
  });
  return modifiedObj;
};

export function omitFieldsFromSchema<T, K extends string | number | symbol>(
  schema: JSONSchemaType<T>,
  fieldsToOmit: (keyof T)[],
): JSONSchemaType<Omit<T, K>> {
  const filteredProperties = removeAttributesFromObject<T, K>(
    schema,
    fieldsToOmit,
  )
  const filteredRequired = schema.required.filter(
    (item: keyof T) => !fieldsToOmit.includes(item),
  )
  return {
    type: 'schema.type',
    properties: filteredProperties,
    required: filteredRequired,
  }
};
```

Reply:
All this is to say, redundant is ugly but I believe it's the best we can do.

The silver lining is that it's reasonable to me that we define the schema explicitly for each API handler anyway, since at the end of the day API signatures should be explicit and hard to accidentally change.

Reply:
I am trying to think if there is ever a second place (outside of the `POST /blah` code) where we would want to validate a minimal or not-yet-inserted entity. I suppose when the situation actually arises, then I would want to put this into a common code area next to the other Ajv types rather than one-off it here. But I agree if it's the only situation then the one-off should be in the context of the one-off.